### PR TITLE
Add LSTM model

### PR DESCRIPTION
**Pull request by yaozengwei**
This PR implements LSTM model. See https://github.com/k2-fsa/icefall/pull/479 for details.

### CODE REVIEW

File: sherpa/csrc/rnnt_rnn_model.h [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+   */
+  torch::Tensor ForwardJoiner(
+      const torch::Tensor &projected_encoder_out,
+      const torch::Tensor &projected_decoder_out) override;
+  /** Run the joiner.encoder_proj network.
+   *
+   * @param encoder_out  The output from the encoder, which is of shape (N,T,C).
+   * @return Return a tensor of shape (N, T, joiner_dim).
+   */
+  torch::Tensor ForwardEncoderProj(const torch::Tensor &encoder_out) override;
```
Comment (csukuangfj) [inline sherpa/csrc/rnnt_rnn_model.h:82]:
```suggestion

```

File: sherpa/csrc/rnnt_rnn_model.h [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+
+  /** Run the joiner.decoder_proj network.
+   *
+   * @param decoder_out  The output from the encoder, which is of shape (N,T,C).
+   * @return Return a tensor of shape (N, T, joiner_dim).
+   */
+  torch::Tensor ForwardDecoderProj(const torch::Tensor &decoder_out) override;
```
Comment (csukuangfj) [inline sherpa/csrc/rnnt_rnn_model.h:89]:
```suggestion

```

File: sherpa/csrc/rnnt_rnn_model.h [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+
+  /** Run the joiner network.
+   *
+   * @param encoder_out  A 2-D tensor of shape (N, C).
+   * @param decoder_out  A 2-D tensor of shape (N, C).
+   * @return Return a tensor of shape (N, vocab_size)
+   */
+  torch::Tensor ForwardJoiner(
+      const torch::Tensor &projected_encoder_out,
+      const torch::Tensor &projected_decoder_out) override;
```
Comment (csukuangfj) [inline sherpa/csrc/rnnt_rnn_model.h:76]:
```suggestion
  torch::Tensor ForwardJoiner(
      const torch::Tensor &encoder_out,
      const torch::Tensor &decoder_out) override;
```

File: sherpa/csrc/rnnt_rnn_model.h [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+
+ private:
+  torch::jit::Module model_;
+
+  // the following modules are just aliases to modules in model_
+  torch::jit::Module encoder_;
+  torch::jit::Module decoder_;
+  torch::jit::Module joiner_;
+  torch::jit::Module encoder_proj_;
+  torch::jit::Module decoder_proj_;
```
Comment (csukuangfj) [inline sherpa/csrc/rnnt_rnn_model.h:108]:
```suggestion

```

File: sherpa/bin/lstm_transducer_stateless/beam_search.py [hunk 5]
```diff
@@ -0,0 +1,316 @@
...
+    def get_texts(self, stream: Stream) -> str:
+        """
+        Return text after decoding
+        Args:
+          stream:
+            Stream to be processed.
+        """
+        return self.sp.decode(
+            stream.hyp[self.beam_search_params["context_size"] :]
+        )
```
Comment (csukuangfj) [inline sherpa/bin/lstm_transducer_stateless/beam_search.py:316]: Please also add

https://github.com/k2-fsa/sherpa/blob/6973dc10ddb47057f398c3c4767e1096cae9a108/sherpa/bin/pruned_stateless_emformer_rnnt2/beam_search.py#L337

**Review by csukuangfj — CHANGES_REQUESTED** [review]
Please also fix the conflicts and integrate the changes from
https://github.com/k2-fsa/sherpa/pull/107

File: .github/workflows/run-streaming-lstm-test.yaml [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+    if: github.event.label.name == 'ready' || github.event_name == 'push'
+    runs-on: ${{ matrix.os }}
+    strategy:
+      fail-fast: false
+      matrix:
+        os: [ubuntu-18.04, macos-latest]
+        torch: ["1.10.0"]
+        torchaudio: ["0.10.0"]
+        python-version: ["3.7", "3.8"]
+        decoding: ["greedy_search", "fast_beam_search", "fast_beam_search_nbest", "fast_beam_search_nbest_LG"]
```
Comment (csukuangfj) [inline .github/workflows/run-streaming-lstm-test.yaml:37]: please also add modified_beam_search.

Comment (yaozengwei) [reply .github/workflows/run-streaming-lstm-test.yaml:37]: OK

**Review by csukuangfj — APPROVED** [review]

**Pull request merged**